helix-term/src/ui/editor.rs

```diff
@@ -746,29 +595,36 @@ impl EditorView {
 
         let gutter_style = theme.get("ui.gutter");
         let gutter_selected_style = theme.get("ui.gutter.selected");
-
-        // avoid lots of small allocations by reusing a text buffer for each line
-        let mut text = String::with_capacity(8);
+        let gutter_style_virtual = theme.get("ui.gutter.virtual");
+        let gutter_selected_style_virtual = theme.get("ui.gutter.selected.virtual");
 
         for gutter_type in view.gutters() {
             let mut gutter = gutter_type.style(editor, doc, view, theme, is_focused);
             let width = gutter_type.width(view, doc);
-            text.reserve(width); // ensure there's enough space for the gutter
-            for (i, line) in (view.offset.row..(last_line + 1)).enumerate() {
-                let selected = cursors.contains(&line);
+            // avoid lots of small allocations by reusing a text buffer for each line
```

Comment:
It's a bit subtle here but `text` actually is re-used for all lines. The boxed closures i  called for each line and since `text` is declared outside of the closure body here it is shared between all calls to the closure which is shared between captures. This example demonstrates this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0d2fdf72e0b329aefbdb686ae2d6290b.